daml-foundations/daml-ghc/src/DA/Daml/GHC/Compiler/Convert.hs

```diff
@@ -569,7 +569,22 @@ convertExpr env0 e = do
             , retrieveByKeyKey = EVar varV1
             }
     go env (VarIs "primitive") (LType (isStrLitTy -> Just y) : LType t : args)
-        = fmap (, args) $ convertPrim (envLfVersion env) (unpackFS y) <$> convertType env t
```

Comment:
This code, and where it lives, is a bit vomit inducing - it takes our beautiful primitive conversion and screws with it.

Any chance we could make the structural records in DAML-LF map type first class, perhaps injected in GHC.Prim like we do for other first-class things, then write a mapping in TextMap module itself that replaces them with tuples?

Reply:
Thanks for pushing me here. I can definitely do better than this. There are also some benefits to be reaped from my new way of doing things regarding contract keys.